sql: replace qvalue with IndexedVar.

knz opened the pull request:
The qvalue mechanism in the sql package was redundant with the
IndexVar mechanism in the sql/parser package. Since the latter is more
general and is already reused in distsql, reuse it in sql and get rid
of qvalue.

Requested by @irfansharif, suggested by @RaduBerinde.
Follows up to (and may fix) #7365.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9821)

<!-- Reviewable:end -->

knz committed:
sql: replace qvalue with IndexedVar.

The qvalue mechanism in the sql package was redundant with the
IndexVar mechanism in the sql/parser package. Since the latter is more
general and is already reused in distsql, reuse it in sql and get rid
of qvalue.

knz commented:
cc @nvanbenschoten

tamird commented:
Reviewed 31 of 31 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[sql/analyze_test.go, line 64 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYAlaEzQ3ttjD62fnx:-KTYAlaEzQ3ttjD62fny:b5fdgu8) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/analyze_test.go#L64)):*
> ```Go
> }
> 
> func resetForTest(sel *selectNode) {
> ```

"reset"? this is already test code.

---

*[sql/check.go, line 89 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYAtksPQHP6ttjIO8f:-KTYAtksPQHP6ttjIO8g:b-9bdcnm) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/check.go#L89)):*
> ```Go
> }
> 
> // IndexedVarEval implements the parser.IndexedVarContainer interface.
> ```

FYI these comments are not needed because checkHelper is not exported.

---

*[sql/data_source.go, line 576 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYB8C98FUheuSfQV0G:-KTYB8C98FUheuSfQV0H:b-v8078a) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/data_source.go#L576)):*
> ```Go
> func (sources multiSourceInfo) findColumn(
> 	c *parser.ColumnItem,
> ) (sIdx int, colIdx int, err error) {
> ```

can you give this a better name than "sIdx"? looks like the method comment is out of date as well.

---

*[sql/returning.go, line 44 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYBpLJS_joyvWHu7UY:-KTYBpLJS_joyvWHu7UZ:bza9h0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/returning.go#L44)):*
> ```Go
> 	alias string,
> 	tablecols []sqlbase.ColumnDescriptor,
> ) (*returningHelper, error) {
> ```

can this avoid changing the return type? i see that `*returningHelper` implements a new interface, but it's not clear that it is always used as that interface, so I'd rather this was left as-is.

---

*[sql/select.go, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYCBFYZI3U2N91X2E9:-KTYCBFYZI3U2N91X2EA:b-vezs72) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/select.go#L48)):*
> ```Go
> 	sourceInfo multiSourceInfo
> 
> 	// Helper for indexed vars. This holds
> ```

weird wrap

---

*[sql/select_name_resolution_test.go, line 59 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYDOE3cx8hrQ7ngUVz:-KTYDOE3cx8hrQ7ngUW-:b-hj9lov) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/select_name_resolution_test.go#L59)):*
> ```Go
> 
> 	for i := 0; i < 2; i++ {
> ```

spurious line

---

*[sql/select_name_resolution_test.go, line 75 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYDYOLSZdT2R5eqPWT:-KTYDYOLSZdT2R5eqPWU:bwpfhri) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/select_name_resolution_test.go#L75)):*
> ```Go
> 		}
> 		if ivars[0].Idx != 0 {
> 			t.Fatalf("%d: ivar not properly initialized for column 0 (a)", i)
> ```

include the unexpected value

---

*[sql/update.go, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYCcbbzTowvAysoHdC:-KTYCcbbzTowvAysoHdD:beyfrox) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/update.go#L37)):*
> ```Go
> type editNodeBase struct {
> 	p          *planner
> 	rh         *returningHelper
> ```

is it now expected for `returningHelper` to be shared?

---

*[sql/parser/indexed_vars.go, line 85 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYCwxlRYLVVRGRdh3R:-KTYCwxlRYLVVRGRdh3S:b-n6xsf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/parser/indexed_vars.go#L85)):*
> ```Go
> func (h *IndexedVarHelper) AssertSameContainer(ivar *IndexedVar) {
> 	if ivar.container != h.container {
> 		panic(fmt.Sprintf("indexed var already resolved with different resolver: %q", ivar.container))
> ```

nit: seems like this message could include a bit more detail

---

*[sql/parser/indexed_vars.go, line 125 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYDLmTTP_6tgu_knTJ:-KTYDLmUbpjjiVDuYfEX:b-v6cl46) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/parser/indexed_vars.go#L125)):*
> ```Go
> 	for i := range h.vars {
> 		if h.vars[i].container == nil {
> 			h.vars[i].Idx = -1
> ```

s/-1/invalidColIdx/?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[sql/analyze_test.go, line 64 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYAlaEzQ3ttjD62fnx:-KTZbbkbwBKQxDhGSpKl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/analyze_test.go#L64)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "reset"? this is already test code.
</details>
Done.

---

*[sql/check.go, line 89 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYAtksPQHP6ttjIO8f:-KTZbct3jegBxIP7fnkZ:b-yxelo0) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/check.go#L89)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> FYI these comments are not needed because checkHelper is not exported.
</details>
Thanks for highlighting; nevertheless I find it useful when I find small methods like these without much context to know which interface requires them.

---

*[sql/data_source.go, line 576 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYB8C98FUheuSfQV0G:-KTZdZ8KnO3oOFVg17dY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/data_source.go#L576)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you give this a better name than "sIdx"? looks like the method comment is out of date as well.
</details>
Done.

---

*[sql/returning.go, line 44 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYBpLJS_joyvWHu7UY:-KTZd_mCJvzoUfb5U7TH:byshhrn) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/returning.go#L44)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can this avoid changing the return type? i see that `*returningHelper` implements a new interface, but it's not clear that it is always used as that interface, so I'd rather this was left as-is.
</details>
Clarified in comments.

---

*[sql/select.go, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYCBFYZI3U2N91X2E9:-KTZeIBtpLBGS9evXzvw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/select.go#L48)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> weird wrap
</details>
Done.

---

*[sql/select_name_resolution_test.go, line 59 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYDOE3cx8hrQ7ngUVz:-KTZeQtZZ5Q5BzQirMmC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/select_name_resolution_test.go#L59)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> spurious line
</details>
Done.

---

*[sql/select_name_resolution_test.go, line 75 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYDYOLSZdT2R5eqPWT:-KTZefgztSdu7nifJP5b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/select_name_resolution_test.go#L75)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> include the unexpected value
</details>
Done.

---

*[sql/update.go, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYCcbbzTowvAysoHdC:-KTZehMbvDkpwSlmonCk:b4b42h5) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/update.go#L37)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is it now expected for `returningHelper` to be shared?
</details>
Yes, as is (now) explained in the comment on newReturningHelper() (see above).

---

*[sql/parser/indexed_vars.go, line 85 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYCwxlRYLVVRGRdh3R:-KTZfBd1hjJCzvNpT7yC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/parser/indexed_vars.go#L85)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: seems like this message could include a bit more detail
</details>
Done.

---

*[sql/parser/indexed_vars.go, line 125 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYDLmTTP_6tgu_knTJ:-KTZfCVZKRiCVpUlzAn-:bemjqnv) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/parser/indexed_vars.go#L125)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/-1/invalidColIdx/?
</details>
invalidColIdx is a member of the sql package, not sql/parser. I'm not sure what the best approach is here. Perhaps migrate these invalid constants to sql/parser and export them?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 31 files at r1, 13 of 13 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[sql/data_source.go, line 576 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYB8C98FUheuSfQV0G:-KT_VwNOX_dDvyzamgeg:bo5d8ed) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/data_source.go#L576)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Done.
</details>
method comment is still wrong, though.

---

*[sql/parser/indexed_vars.go, line 125 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYDLmTTP_6tgu_knTJ:-KT_Vls-IgoXTOorJbPb:b4r6fgs) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/parser/indexed_vars.go#L125)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> invalidColIdx is a member of the sql package, not sql/parser. I'm not sure what the best approach is here. Perhaps migrate these invalid constants to sql/parser and export them?
</details>
Yeah, that might be the way to do it.

---

*[sql/parser/star_datum.go, line 50 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KT_VpEggZIHavy-MWvm:-KT_VpEggZIHavy-MWvn:bfxz830) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa6f617678673c82f53506413edce042b992e0cb/sql/parser/star_datum.go#L50)):*
> ```Go
> 
> // TypeCheck implements the Expr interface.
> func (s *StarDatum) TypeCheck(_ *SemaContext, desired Datum) (TypedExpr, error) {
> ```

s/desired/_/ while you're here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 26 of 30 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/data_source.go, line 576 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYB8C98FUheuSfQV0G:-KT_hfRG7Gtk39xNtd-k:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/data_source.go#L576)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> method comment is still wrong, though.
</details>
Done.

---

*[sql/parser/indexed_vars.go, line 125 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTYDLmTTP_6tgu_knTJ:-KT_hibrbLofCcsHm16c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/245de34817352620626f18b43cc385ef3b1d6d77/sql/parser/indexed_vars.go#L125)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Yeah, that might be the way to do it.
</details>
Done.

---

*[sql/parser/star_datum.go, line 50 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KT_VpEggZIHavy-MWvm:-KT_hhwSP3yt6Ru2-DjK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa6f617678673c82f53506413edce042b992e0cb/sql/parser/star_datum.go#L50)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/desired/_/ while you're here
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but you should have someone who's slightly less :dog: take a look.

---

Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/select_name_resolution.go, line 29 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KT_mE1FaZIvB__P_Ew2:-KT_mE1FaZIvB__P_Ew3:b-9aabm7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/select_name_resolution.go#L29)):*
> ```Go
> 
> const invalidSrcIdx = -1
> const invalidColIdx = parser.InvalidColIdx
> ```

worth keeping both? just a simple search & replace, no?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-:-KT_mOc458m6-_3XMVRa:b-4dpa5o)*
<!-- Sent from Reviewable.io -->

knz commented:
Yes, good idea. Since Radu invented IndexedVar I hope to get some feedback from him too. TFYR nonetheless.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/select_name_resolution.go, line 29 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KT_mE1FaZIvB__P_Ew2:-KTa5iopKO15mt6Hk8kW:b-2ig88c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/select_name_resolution.go#L29)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> worth keeping both? just a simple search & replace, no?
</details>
I don't really want to read "invalidSrcIdx" and "parser.InvalidColIdx" side-by-side in this package.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-:-KTa5brD0K6uR8xJzlLP:bgmgdzx)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/select_name_resolution.go, line 29 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KT_mE1FaZIvB__P_Ew2:-KTaA12pu0A6tCgRJihg:bvt9joe) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/select_name_resolution.go#L29)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I don't really want to read "invalidSrcIdx" and "parser.InvalidColIdx" side-by-side in this package.
</details>
I suppose I was imagining you'd move `invalidSrcIdx` as well, but this is fine too. A small comment wouldn't hurt.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
This sounds awesome, I will definitely take a look!

RaduBerinde commented:
The `sql/parser/...` part looks good.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[sql/parser/indexed_vars.go, line 83 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTdW8_WUpMsNoUGt0GL:-KTdW8_WUpMsNoUGt0GM:b-83sjoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/parser/indexed_vars.go#L83)):*
> ```Go
> 
> // AssertSameContainer checks that the indexed var refers to the same container.
> func (h *IndexedVarHelper) AssertSameContainer(ivar *IndexedVar) {
> ```

Nice!

---

*[sql/parser/indexed_vars.go, line 126 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTdWVBVm-m7eMG7zp1u:-KTdWVBVm-m7eMG7zp1v:b-fu9nlz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/parser/indexed_vars.go#L126)):*
> ```Go
> // GetIndexedVars retrieves the array of initialized IndexedVars;
> // unused vars are guaranteed to have an invalid index.
> func (h *IndexedVarHelper) GetIndexedVars() []IndexedVar {
> ```

We are returning a direct reference to an internal slice. Should we document some restrictions on the caller? (we probably want to disallow any modification)

It might be cleaner to use a `GetColIdx(varIdx int) int` instead of a getter for the entire array.

---

*[sql/parser/star_datum.go, line 28 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTdY4TScz443XQIPqji:-KTdY4TTnJNTgNra-LoW:bmgmo4z) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/parser/star_datum.go#L28)):*
> ```Go
> // allow the intermediate rendering of the value (before the group
> // aggregation).
> type StarDatum struct{}
> ```

Can you add a mention that this is not to be used for other cases of `*`, like when `*` refers to a set of columns? (it's kind of clear from the name `Datum`, but still)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-:-KTdYMze5iibBzIkZanV:b-6f2vjx)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>
 
Only minor comments.

---

Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[sql/analyze_test.go, line 140 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTdlHmruUr1R54-b2hO:-KTdlHmruUr1R54-b2hP:bz901cs) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/analyze_test.go#L140)):*
> ```Go
> 	for _, d := range testData {
> 		t.Run(d.expr+"~"+d.expected, func(t *testing.T) {
> 			reset(sel)
> ```

why don't we just do `sel := makeSelectNode()` here? It's kind of hard to think about these tests sharing the pointer. (or maybe init the `desc` once and do `testInitDummySelectNode` in each test time).
If we don't have the same pointer, we could also run `t.Parallel()` to run the cases in parallel for faster test runs.

---

*[sql/analyze_test.go, line 275 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTdpQ7kwaIolDkNZbAm:-KTdpQ7kwaIolDkNZbAn:b-18pn0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/analyze_test.go#L275)):*
> ```Go
> 	}
> 	for _, d := range testData {
> 		t.Run(d.expr+"~"+d.expected, func(t *testing.T) {
> ```

These test names are going to be insane.. They may cause various issues (e.g. when automatic issues are filed).

---

*[sql/returning.go, line 43 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTdrn1b2YFeAGg2B4dx:-KTdrn1cD1d9XKkf-RLl:b-uomoq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/returning.go#L43)):*
> ```Go
> // its location stay stable after construction because it implements
> // IndexedVarContainer and the IndexedVar objects in sub-expressions
> // will link to it by reference after checkRenderStar / analyzeExpr.
> ```

Consider embedding a `noCopy` struct to `returningHelper` and moving this comment there. It may help catch bugs early (Irfan ran into something like this and it was hard to track down. though your AssertSameContainer will already help a lot).

https://github.com/cockroachdb/cockroach/blob/1432e7be3c7efb4601a0d33f961ca392ad0805a7/sql/distsql/expr.go#L144

---

*[sql/select_name_resolution.go, line 148 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTdjKLFeRVLR3tphn_7:-KTdjKLFeRVLR3tphn_8:bywwa7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/select_name_resolution.go#L148)):*
> ```Go
> ) (parser.Expr, error) {
> 	if expr == nil {
> 		return expr, nil
> ```

[nit] nil, nil

---

*[sql/select_name_resolution.go, line 153 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTdkW3Z8DwckQNhJsrj:-KTdkW3Z8DwckQNhJsrk:b-q59kls) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/select_name_resolution.go#L153)):*
> ```Go
> 		v = new(nameResolutionVisitor)
> 	}
> 	*v = nameResolutionVisitor{
> ```

Can't we make `v` a local `nameResolutionVisitor` and just pass `&v` to `WalkExpr`? (I'm assuming go figures out that it doesn't escape the `WalkExpr` call but I may be optimistic)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-:-KTdsNHre6rnQN6p4aqR:b-3qtxry)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[sql/analyze_test.go, line 275 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTdpQ7kwaIolDkNZbAm:-KTduAnsq2lQD3j9tWIu:bvnt4e3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/analyze_test.go#L275)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> These test names are going to be insane.. They may cause various issues (e.g. when automatic issues are filed).
</details>
Actually, disregard this. I see this exact pattern is shown on the Go blog https://blog.golang.org/subtests ("Table-driven tests using subtests"). If there are any infrastructure issues with weird test names we will just have to fix them.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821)*
<!-- Sent from Reviewable.io -->

knz committed:
sql: fix TestLogic.

Previous commit 4ce9b6efc9f2cea52cbce5bde55c9c2e60b4c861 has
introduced sub-tests in TestLogic. However TestLogic accesses
testing.T via the logicTest wrapper struct, and this is initialized
outside of the sub-test Run call. Prior to this patch, calling
l.Fatal() on the logicTest would call t.Fatal on the outer testing.T
instance, not the one for the sub-test; this causes any
error in the sub-test to be lost upon Fatal().

This patch fixes the issue by ensuring that the logicTest wrapper
struct wraps the testing.T for the sub-test instead.

knz committed:
util: new helper struct NoCopy.

knz commented:
Review status: 0 of 33 files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/analyze_test.go, line 140 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTdlHmruUr1R54-b2hO:-KTfiZ_j6mUeqkHFzsTB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/analyze_test.go#L140)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> why don't we just do `sel := makeSelectNode()` here? It's kind of hard to think about these tests sharing the pointer. (or maybe init the `desc` once and do `testInitDummySelectNode` in each test time).
> If we don't have the same pointer, we could also run `t.Parallel()` to run the cases in parallel for faster test runs.
</details>
Done.

---

*[sql/returning.go, line 43 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTdrn1b2YFeAGg2B4dx:-KTgF1boDkO7Qt8Jn_zQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/returning.go#L43)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Consider embedding a `noCopy` struct to `returningHelper` and moving this comment there. It may help catch bugs early (Irfan ran into something like this and it was hard to track down. though your AssertSameContainer will already help a lot).
> 
> https://github.com/cockroachdb/cockroach/blob/1432e7be3c7efb4601a0d33f961ca392ad0805a7/sql/distsql/expr.go#L144
</details>
Done.

---

*[sql/select_name_resolution.go, line 29 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KT_mE1FaZIvB__P_Ew2:-KTfk8pUpGO1ZKn1FO2a:b76ucq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/select_name_resolution.go#L29)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I suppose I was imagining you'd move `invalidSrcIdx` as well, but this is fine too. A small comment wouldn't hurt.
</details>
I added a comment. The srcIdx cannot be conceptually part of the sql/parser package because the notion of data source does not exist at all in there.

---

*[sql/select_name_resolution.go, line 148 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTdjKLFeRVLR3tphn_7:-KTfkJIS3cRnM275Qthj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/select_name_resolution.go#L148)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> [nit] nil, nil
</details>
Done.

---

*[sql/select_name_resolution.go, line 153 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTdkW3Z8DwckQNhJsrj:-KTfmCrogt_NGOQTJgS7:b-puhi8i) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/select_name_resolution.go#L153)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Can't we make `v` a local `nameResolutionVisitor` and just pass `&v` to `WalkExpr`? (I'm assuming go figures out that it doesn't escape the `WalkExpr` call but I may be optimistic)
</details>
Done (in a different way though. Much cleaner.)

---

*[sql/parser/indexed_vars.go, line 126 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTdWVBVm-m7eMG7zp1u:-KTfnnDkrR-J2r3ED5nU:bn4h8xh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/parser/indexed_vars.go#L126)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> We are returning a direct reference to an internal slice. Should we document some restrictions on the caller? (we probably want to disallow any modification)
> 
> It might be cleaner to use a `GetColIdx(varIdx int) int` instead of a getter for the entire array.
</details>
The thing is the callers want the array but do not need to hold on the helper. I have changed the interface to simply transfer the ownership entirely.

---

*[sql/parser/star_datum.go, line 28 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTdY4TScz443XQIPqji:-KTfmRfmRvqas3jUG-Dj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f67beb7f2cd23c8b90a890cc9ce05e5a69bf8b3b/sql/parser/star_datum.go#L28)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Can you add a mention that this is not to be used for other cases of `*`, like when `*` refers to a set of columns? (it's kind of clear from the name `Datum`, but still)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 31 files at r1, 31 of 32 files at r4, 1 of 1 files at r5, 18 of 31 files at r6, 1 of 33 files at r7, 1 of 1 files at r8, 31 of 31 files at r9.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[build/check-style.sh, line 162 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTgJixvivi7WkmrgPUp:-KTgJixvivi7WkmrgPUq:b5dd6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2be3b667d72ccf9b4b28d588ff9d206f9f7225a/build/check-style.sh#L162)):*
> ```Shell
> 
> TestUnused() {
>   ! ((unused -reflect=false -exported ./... || true) | grep -vE 'sql/(pgwire/pgerror/codes.go|parser/yacc(par|tab))|(field|type) [nN]oCopy ')
> ```

perhaps this should be `(field no|type No)Copy`

---

*[sql/select.go, line 94 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTgJ1hWDPx_aISDkl_J:-KTgJ1hWDPx_aISDkl_K:b-bjinnk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5c7747768750565cf53c62d527d2bcaf8b34632/sql/select.go#L94)):*
> ```Go
> 	row parser.DTuple
> 
> 	// Each selectNode must be allocated on the heap and its
> ```

all these comments mention the structure they're in by name, which is prone to rot. up to you, but consider avoiding that as much as possible without sacrificing clarity.

---

*[sql/parser/indexed_vars.go, line 141 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTgJYvD167tHYCmaHsv:-KTgJYvD167tHYCmaHsw:b6is663) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5c7747768750565cf53c62d527d2bcaf8b34632/sql/parser/indexed_vars.go#L141)):*
> ```Go
> 	}
> 	ret := h.vars
> 	//	h.vars = nil
> ```

did you mean to leave this commented out?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 1 of 33 files reviewed at latest revision, 11 unresolved discussions.

---

*[build/check-style.sh, line 162 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTgJixvivi7WkmrgPUp:-KTgKdBqDKHhqLFiLZZS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2be3b667d72ccf9b4b28d588ff9d206f9f7225a/build/check-style.sh#L162)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> perhaps this should be `(field no|type No)Copy`
</details>
Done.

---

*[sql/select.go, line 94 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTgJ1hWDPx_aISDkl_J:-KTgN_mwOhlFzEFEWaiz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5c7747768750565cf53c62d527d2bcaf8b34632/sql/select.go#L94)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> all these comments mention the structure they're in by name, which is prone to rot. up to you, but consider avoiding that as much as possible without sacrificing clarity.
</details>
Done.

---

*[sql/parser/indexed_vars.go, line 141 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KTgJYvD167tHYCmaHsv:-KTgK_GVS-KTrYMddfMc:b-d3ketw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5c7747768750565cf53c62d527d2bcaf8b34632/sql/parser/indexed_vars.go#L141)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> did you mean to leave this commented out?
</details>
woops, no, oversight. Good catch!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[sql/parser/indexed_vars.go, line 127 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KThOicqhHDusjZzxCYr:-KTiv65iKErsrWNeT5NZ:bi5lynp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2be3b667d72ccf9b4b28d588ff9d206f9f7225a/sql/parser/indexed_vars.go#L127)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how come this changed?
</details>
What has changed? The test on `container` is sufficient.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[sql/parser/indexed_vars.go, line 127 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KThOicqhHDusjZzxCYr:-KTjIpbCZwqZY2I9TKwU:b-yq4hm2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2be3b667d72ccf9b4b28d588ff9d206f9f7225a/sql/parser/indexed_vars.go#L127)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> What has changed? The test on `container` is sufficient.
</details>
If you click on the `r9` at the top right corner of this thread box, you'll be shown the diff I saw when I left that comment.

If the test on container is sufficient then this is fine (though i don't understand why there's no need to check against InvalidColdx).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[sql/parser/indexed_vars.go, line 127 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821#-KThOicqhHDusjZzxCYr:-KTjSBSeARCjJCza4-up:burnl5x) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2be3b667d72ccf9b4b28d588ff9d206f9f7225a/sql/parser/indexed_vars.go#L127)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> If you click on the `r9` at the top right corner of this thread box, you'll be shown the diff I saw when I left that comment.
> 
> If the test on container is sufficient then this is fine (though i don't understand why there's no need to check against InvalidColdx).
</details>
r9 was an intermediate state with extra checks for debugging, I had forgotten to revert the change (as you found out earlier).

The reason why the check on container is sufficient is that IndexedVar users can only get a new IndexedVar from an IndexedVarHelper using the IndexedVar() method, which sets both the container and the index together.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9821)*
<!-- Sent from Reviewable.io -->
